.gitmodules

```diff
@@ -10,3 +10,7 @@
 [submodule "thirdparty/libbcrypt"]
 	path = thirdparty/libbcrypt
 	url = https://github.com/trusch/libbcrypt.git
+# TODO: Merge changes by TheNoim into lcdr's repo and use that
```

Comment:
I wouldn't advise merging this with a reference to a fork and thus far no PR to the utils has been created.

Also the changes in question are a CLI pkextractor. I actually have something like that ready to go but not public yet.

Reply:
lcdr specifically said the utils were always supposed to be w/ GUI in this issue I made https://github.com/lcdr/utils/issues/3. It doesn't sound like he has intentions to accept a PR to his main repo

Reply:
That assumption is likely correct, but missing some context. My comment
> I actually have something like that ready to go but not public yet.
and @lcdr's comment
> I myself think it's unnecessarily complicated right now. Anyway, the DLU team is working on making it easier as we speak.
essentially refer to the same thing to my knowledge:

1. The DLU server doesn't need a full extraction. While PKExtractor can do single file, this PR seems to use the complete extraction. The CLI tool I'm referring to is ~200 lines of code on top of <https://github.com/xiphoseer/assembly_rs> can take a set of globs (e.g. `client/res/BrickModels/**`) as input and extract just those files.
2. PKExtractor is very old, older than much of our understanding on how the game uses the PK files. It scans for a specific hash from trunk.txt, which can add a lot of overhead, especially if you don't need to extract all files anyways. We now know that every path can be converted to a number (CRC) and be used with the PKI file to directly identify which PK file holds the data, and, to find the data within in the PK file. That's what "unnecessarily complicated" refers to.
3. I actually chatted with @lcdr about this and know for a fact that he thinks "[my] tools could improve the setup workflow quite a bit".
4. A lot of random people trust this repo enough to download what's here and just run it on their machine. There's a well defined group of people that can merge code into this repo. If we include a fork as a submodule, we would pass on that trust to someone else. In my opinion, we should avoid that or at least be cautious and explicit about that.

Reply:
PS: Feel free to DM me on Discord to get a preview of the tool I'm talking about.

Reply:
A couple of notes from my perspective:

> I wouldn't advise merging this with a reference to a fork

I agree with this, adding dependencies to outside sources could be a security issue and should be avoided.

> lcdr specifically said the utils were always supposed to be w/ GUI in this issue I made lcdr/utils#3. It doesn't sound like he has intentions to accept a PR to his main repo

That's not quite correct, I said the specific tool in question (pkextractor) was meant to be a GUI app, and I said that as an explanation why it doesn't have a CLI mode right now, not as a defining feature. There are other tools in the repo that are CLI tools. I would be happy to merge a PR that either adds a separate CLI tool or adds CLI support to pkextractor, it's just that no one's done it so far :P

I haven't seen the fork in question, but I'm happy to hear that someone's forked my repo and made improvements. Make a PR to the utils repo and we can see what can be done to get it merged.

> That assumption is likely correct, but missing some context. My comment
> > I actually have something like that ready to go but not public yet.

> and @lcdr's comment
> > I myself think it's unnecessarily complicated right now. Anyway, the DLU team is working on making it easier as we speak.

> essentially refer to the same thing to my knowledge

Actually, I was referring to the efforts by Mick, Wincent, and a couple others to provide a python script that automates the setup workflow (not sure of the status of that), not something specific to PKs, but I definitely also agree in particular that the PK extraction step could and should be automated.

> The DLU server doesn't need a full extraction. While PKExtractor can do single file, this PR seems to use the complete extraction. 

I agree with this, a tool used to automate the setup should only extract what's needed, as there'd be significant inefficiencies otherwise.

> I actually chatted with @lcdr about this and know for a fact that he thinks "[my] tools could improve the setup workflow quite a bit".

Well, I meant that your tools (as programmatic / CLI access to PK files) would be better for the DLU setup than my GUI, as they could be automated. I wasn't aware of this fork at that time. I'd just like to clarify that I have no preferences for specific implementations, just of a CLI over a GUI for the setup usecase. Whether the setup ends up using Python-based tools or Xipho's Rust tools doesn't matter to me, as long as the features (CLI, efficient partial extraction, etc) are suitable to the setup workflow. It's up the PR author to decide whether to PR changes to my utils and reference them in the setup, or to integrate Xipho's tools instead, I'd be fine with both. (And a PR to the utils repo would be appreciated in any case. 😄 )

Reply:
Sorry, I misunderstood. I agree with the points you both made. It makes sense to figure out how to handle this before merging the PR.

Reply:
I agree with that we should change/replace this tool later, but at this moment I don't see this as a road blocker for this PR. The submodule is locked to a specific commit. This basically means, after this pr got merged, I am not able to change anything in my fork which could harm the end user and this is how it should be. If we later decide to use a better written tool which could replace PKExtractor a new pr needs to happen anyway. Currently, I don't see a reason to make a PR to @lcdr repo, because I basically took his tool and removed large chunks of it (the GUI part). I understand that his GUI tool is useful for users, but it is just not required for this PR to Darkflame. Of course, if @lcdr also really wants this changes to his fork, I can prepare a pr, but at the moment I don't see the usefulness of investing the time in such a PR.

If you all really don't feel comfortable me having the ownership of the fork, I could transfer the ownership of my it to Darkflame, and they could have a protected master branch and I can work on a separate one. The same thing then also needs to happen with the newly developed version from @Xiphoseer based on rust (btw I am very excited to see a finished version which could replace PKExtractor).

> > The DLU server doesn't need a full extraction. While PKExtractor can do single file, this PR seems to use the complete extraction.

> I agree with this, a tool used to automate the setup should only extract what's needed, as there'd be significant inefficiencies otherwise.

I am aware of this. But to my current knowledge, there is no list of absolute needed files (If I am wrong, please correct me). I am happy to change this and only extract required files, but at the moment it is just better to extract all.

> PS: Feel free to DM me on Discord to get a preview of the tool I'm talking about.

Whats your Discord tag then?

Reply:
> But to my current knowledge, there is no list of absolute needed files (If I am wrong, please correct me)
`versions/trunk.txt` is a sufficiently complete list of files and my tools uses a globset file such as
```txt
client/res/macros/**
client/res/BrickModels/**
client/res/maps/**
```
to narrow that down to the subset DLU needs. But you can do whatever you need including `*.dds` (for all images) or `*.nif` for all meshes for example

> Whats your Discord tag then?

Xiphoseer#0684 -- Usually just the name is sufficient to find me

Reply:
@TheNoim can you try out using https://github.com/Xiphoseer/LUnpack ?

Reply:
Awesome!